KAFKA-18401: Transaction version 2 does not support commit transaction without records

brandboat opened the pull request:
related to KAFKA-18401, 

Fix the issue where producer.commitTransaction under transaction version 2 throws error if no partition or offset is added to transaction. The solution is to avoid sending the endTxnRequest unless producer.send or producer.sendOffsetsToTransaction is triggered.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

brandboat committed:
KAFKA-18401: Transaction version 2 does not support commit transaction without records

jolshan commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -854,7 +856,7 @@ synchronized TxnRequestHandler nextRequest(boolean hasIncompleteBatches) {
         }
 
-        if (nextRequestHandler.isEndTxn() && (!isTransactionV2Enabled() && !transactionStarted)) {
+        if (nextRequestHandler.isEndTxn() && !transactionStarted) {
```
minor nit here: I wonder if we can change the log here for tv2 so that we say we didn't attempt to add any partitions.

brandboat replied:
Pardon me, could you give me some hint? Did you mean we need to modify the log in line 862
```java
                log.debug("Not sending EndTxn for completed transaction since no partitions " +
                        "or offsets were successfully added");
```
to something else when using TV2?

jolshan replied:
Right -- the implication of successfully added is a little confusing in the TV2 case because it is not a matter of "successfully" adding but a matter of not attempting to add any at all.

brandboat replied:
Got it, I modified the log to `Not sending EndTxn for completed transaction since no send or sendOffsetsToTransaction were triggered` under TV2. Thank you.

jolshan commented on core/src/test/java/kafka/clients/producer/ProducerIntegrationTest.java:
```diff
@@ -0,0 +1,55 @@
+        @ClusterTest(brokers = 3, features = {
+            @ClusterFeature(feature = Feature.TRANSACTION_VERSION, version = 2)})
+    })
+    public void testTransactionWithoutSend(ClusterInstance cluster) {
```
should we include tests for when we do send/add a partition to confirm we do see errors?

jolshan replied:
And is it possible to add these tests to an existing file?

brandboat replied:
> should we include tests for when we do send/add a partition to confirm we do see errors?

Sure, I'm going to write one!

> And is it possible to add these tests to an existing file?

Perhaps this one? https://github.com/apache/kafka/blob/trunk/core/src/test/scala/integration/kafka/coordinator/transaction/ProducerIdsIntegrationTest.scala ? We can rename this file to `ProducerIntegrationTest.scala` and move the test to this file. WDYT?

jolshan replied:
That is an option.

brandboat replied:
OK, I'll move the test in this PR to the test file I mentioned above. Thanks!

brandboat committed:
Add more test

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18401

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18401

brandboat committed:
Refine test

brandboat committed:
Refine test

brandboat committed:
Remove unnecessary test and rename tests

brandboat committed:
Address comments

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18401

jolshan commented on core/src/test/scala/integration/kafka/coordinator/transaction/ProducerIntegrationTest.scala:
```diff
@@ -0,0 +1,185 @@
+    new ClusterTest(features = Array(
+      new ClusterFeature(feature = Feature.TRANSACTION_VERSION, version = 2))),
+  ))
+  def testTransactionWithSendOffset(cluster: ClusterInstance): Unit = {
```
Does this test new behavior/one related to the change? The one other thing I was curious about was a test where we try to send a record, it fails to be added and we should be able to abort, but a commit (in the case where a partition was not added) will fail. That one might be tricky to code -- especially the part where we fail to add the partition on the server side.

brandboat replied:
> Does this test new behavior/one related to the change?

Yes, I want to ensure that if a transaction with sendOffsetToTransaction triggered and no sendRecords are fired, then executing commitTransaction should not raise any errors. And on the server side, the transaction state should be COMPLETE_COMMIT, which means the END_TXN request has been sent to the broker.

> The one other thing I was curious about was a test where we try to send a record, it fails to be added and we should be able to abort, but a commit (in the case where a partition was not added) will fail. That one might be tricky to code -- especially the part where we fail to add the partition on the server side.

Hmm... I need some time to figure out how to make adding the partition fail on server side. Once that's done, I think I can come up with the test you mentioned.

Thanks for the suggestion!

brandboat replied:
There is one test, `testFencingOnAddPartitions` in TransactionTest.scala [[0]], which seems to do something similar—creating a scenario where `TransactionCoordinator#handleAddPartitionsToTransaction` raises an error, `Errors.PRODUCER_FENCED` due to a mismatch in the producer epoch [[1]] and cause AddPartitionToTxn request failed.

However, in this test, we cannot commit or abort the transaction successfully. Committing the transaction directly fails from the client side since the last error caused by the failed send is not empty[[2]]. In the case of aborting the transaction, since there is a newer producer (producer2), the broker raises the error Errors.PRODUCER_FENCED, causing the abort to fail.

Perhaps the producer fenced test case is not what we want? I'm just wondering under what circumstances the `AddPartitionToTxn` would fail and in producer we can abort transaction successfully but commit transaction would fail.

[0]: https://github.com/apache/kafka/blob/f79d7dc3f4d8418561f8c89bfa7ad19e529005b8/core/src/test/scala/integration/kafka/api/TransactionsTest.scala#L557
[1]: https://github.com/apache/kafka/blob/3ef10a3f617cfa7dc0bc7cb29dc2e3950765f1dc/core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala#L413-L414
[2]: https://github.com/apache/kafka/blob/94eb21cab4e86d0d777553b0a52af77e32c2f20f/clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java#L319

jolshan replied:
This is a bit tricky. Depending on the error, a producer will enter a fatal error state or an abortable error state. In the abortable error state, we can always abort, but not commit. In fatal state, we can't do anything. The fenced error I believe puts the producer in fatal state. There are a few other errors that are only "abortable" but not sure if they can easily come up in test scenarios. Some there are some auth errors like this for example, but that may be hard to replicate on a test. KIP-1050 does a pretty good job explaining the current state of transactional errors. You can see in the "current handling" columns of the tables in the KIP. https://cwiki.apache.org/confluence/display/KAFKA/KIP-1050:+Consistent+error+handling+for+Transactions

I'm thinking though that this might be a bit complicated for testing, and it is ok to not include if it is too much trouble. I think the main thing I wanted to confirm is that if we attempted to send a record/offset, we can no longer do an empty commit.

brandboat replied:
> I think the main thing I wanted to confirm is that if we attempted to send a record/offset, we can no longer do an empty commit.

Sorry for the delay, I was a little bit confused about this comment, what does an empty commit means? Does `testTransactionWithAndWithoutSend` in ProducerIntegrationTest.scala cover that?

brandboat replied:
Or did you mean something like
```java
      producer.initTransactions()
      producer.beginTransaction()
      producer.send(new ProducerRecord[Array[Byte], Array[Byte]]("test", "key".getBytes, "value".getBytes)).get()

      producer.beginTransaction() // <-- fail TransactionalId foobar: Invalid transition attempted from state IN_TRANSACTION to state IN_TRANSACTION
      producer.commitTransaction()
```
in this case, it would fail in client side actually

jolshan replied:
Sorry for not being clear. I meant something similar to the above, but we wouldn't call begin transaction again. Something like 

```
      producer.initTransactions()
      producer.beginTransaction()
      producer.send(new ProducerRecord[Array[Byte], Array[Byte]]("test", "key".getBytes, "value".getBytes)).get() 
      // we would create a case where this fails and we wouldn't actually write a record. I think it would take us to an abortable state

      producer.commitTransaction() // fails
```

Basically, just wanted to confirm that once we hit "true" for transaction started, we don't skip the EndTxn in the new code that has been added. I guess maybe this is a very specific scenario with the send failing and it is sufficient to just have the test where once we send/add offsets, we confirm the EndTxn sent. (No need for the error.) If we do that, I think your current test may be sufficient.

brandboat replied:
Thanks for the explanation, I think I need to figure out a scenario that  `producer.send(new ProducerRecord[Array[Byte], Array[Byte]]("test", "key".getBytes, "value".getBytes)).get()` failed to send record. Will update the test ASAP!

brandboat replied:
~~Sorry I just found my last update test is still wrong... I'm still using fatal error scenario, will try to find other way, plz ignore the commit [bded873]~~(https://github.com/apache/kafka/pull/18448/commits/bded873f764b9ef4f9295de6bb8a22171b04b8d5)

jolshan commented:
It's looking pretty good @brandboat. I just have one more question about the testing we are doing.

jolshan commented:
@brandboat Any updates here? Did we want to add/change any tests? This was the only thing I was thinking might be remaining

>  I think the main thing I wanted to confirm is that if we attempted to send a record/offset, we can no longer do an empty commit.

brandboat committed:
Address comments

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18401

jolshan commented on core/src/test/scala/integration/kafka/coordinator/transaction/ProducerIntegrationTest.scala:
```diff
@@ -85,6 +88,50 @@ class ProducerIntegrationTest {
+      val commitError = assertThrows(classOf[KafkaException], () => producer.commitTransaction()) // fail due to last send failed
+      assertInstanceOf(classOf[RecordTooLargeException], commitError.getCause)
+
+      if (txnVersion == 2) {
```
nice! happy to see this worked! 
small nit -- is there a reason to run this test on the other versions? I think we should either have cases to test for versions 0 and 1 or not run those for this one. WDYT?

jolshan replied:
Oops. I see your comment now. I will look at the revised version when it is ready, but I think the nit still stands.

brandboat replied:
> small nit -- is there a reason to run this test on the other versions? I think we should either have cases to test for versions 0 and 1 or not run those for this one. WDYT?

Yeah.. though I think the test case perhaps is wrong. But I observed that the abortTransaction in this test case only works under TV_2 instead of TV_0 and TV_1. I haven't look into it yet, not sure why...

brandboat replied:
Oops, in the beginning I thought `RecordTooLargeException` will put transaction in `FATAL_ERROR` state in TransactionManager, but the state is `ABORTABLE_ERROR` actually, so the test seems in the right direction.

I'll deep dive and check why abortTransaction will hang under TV_0 & TV_1 in the test case.

brandboat replied:
Hi @jolshan , I just update the test, now the test seems ok. 

> small nit -- is there a reason to run this test on the other versions? I think we should either have cases to test for versions 0 and 1 or not run those for this one. WDYT?

After I increment max.message.bytes to 100, abortTransaction not hanging anymore with TV_0 and TV_1, but I didn't find out the root cause of why doing this could solve the hanging issue...

jolshan replied:
I think the marker takes up some bytes :) so we need space for that. With TV0 and TV1 we add the partition before even sending (and getting an error) so when we abort we try to still write a marker to that partition.

I'm still working through why the TV2 case was working. Not quite sure yet.

brandboat replied:
> I'm still working through why the TV2 case was working. Not quite sure yet.

~~Looks like in TV_2 we'll call response (EndTxnResponse) first, and then call WriteTxnMarkerRequest, I think that's why TV2 abortTransaction pass even if WriteTxnMarker failed (due to max.message.bytes=1, like you said, marker takes up some bytes)~~ (Update, this behavior is the same as TV_0, TV_1)

https://github.com/apache/kafka/blob/32dbbe6a1f3ef39318c796bdc0a3b8da2c7060ad/core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala#L943-L945

jolshan replied:
This should be the same for TV0 and TV1 though...

brandboat replied:
Oh... should I open another JIRA ticket?

jolshan replied:
I'm not sure I understand what the ticket would be. I will continue to think about this, but I don't think it needs to block this PR.

jolshan replied:
I recreated the failure locally and it seems like for TV0 and TV1 we are unable to join the ioThread when trying to close the producer. With TV2, the thread seems to join fine. Were you observing this as well?

brandboat replied:
Sorry for the late reply, the comment https://github.com/apache/kafka/pull/18448#discussion_r1916916687 was incorrect—apologies for the misleading input. The behavior I was referring to, where the response (EndTxnResponse) is called first, followed by WriteTxnMarkerRequest, is actually the same for TV0 and TV1.

> I recreated the failure locally and it seems like for TV0 and TV1 we are unable to join the ioThread when trying to close the producer. With TV2, the thread seems to join fine. Were you observing this as well?

What I observed in TV0 and TV1 is that the producer continuously sends InitProducerId requests[[0]][[1]] because it encounters Errors.CONCURRENT_TRANSACTIONS. Since this is a retriable error, the producer enters a retry loop, repeatedly sending InitProducerId requests to the broker. But the last transaction remains incomplete, the entire process gets stuck in an endless retry loop, and causing the test time out. 

The CONCURRENT_TRANSACTIONS error occurs because WriteTxnMarker fails due to a RecordTooLargeException, preventing the transaction state from correctly transitioning to the CompleteAbort state.

Although TV2 passes in this scenario, we can still reproduce the CONCURRENT_TRANSACTIONS error if we add the following lines after abortTransaction(). This happens because WriteTxnMarker failed in the previous abort attempt:
```
      producer.beginTransaction()
      producer.send(new ProducerRecord[Array[Byte], Array[Byte]]("test", "key".getBytes, "value".getBytes))
      producer.commitTransaction()
```

I'm not sure if this is something we need to aware of...

[0]: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java#L362
[1]: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java#L1445

jolshan replied:
Let me ponder this a bit. Thanks @brandboat for the deeper investigation. Is this because the TV0, TV1, send intProducerId when it needs to bump the epoch? This is a difference between TV0/1 and TV2. 

If this is correct, the issue is that any request after EndTxn will see concurrent transactions, which makes sense and is expected.

brandboat replied:
> If this is correct, the issue is that any request after EndTxn will see concurrent transactions, which makes sense and is expected.

Yep, that's the root cause, glad to hear that this is not an issue. Thanks for the explanation. 😃

brandboat committed:
Update test

brandboat committed:
Remove listTransactions checks in test

this check makes test flaky, and abortTransaction
will send EndTxnRequest anyway, this check seems
redundant here.

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18401

brandboat committed:
Remove redundant code

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18401

jolshan reviewed (approved):
Thanks for your diligence with the tests @brandboat

brandboat commented:
> Thanks for your diligence with the tests @brandboat

Thanks for your patience, I learned a lot!
